data/blog/libtailscale.mdx

```diff
@@ -0,0 +1,198 @@
+It is built on the Go [tsnet](https://pkg.go.dev/tailscale.com/tsnet)
+package which uses a userland network stack to implement TCP/IP
+directly inside your process.
+
+```
+tailscale ts = tailscale_new();
+if (tailscale_set_ephemeral(ts, 1)) {
+	return err(ts);
+}
+if (tailscale_up(ts)) {
+	return err(ts);
+}
+tailscale_listener ln;
+if (tailscale_listen(ts, "tcp", ":1999", &ln)) {
+	return err(ts);
+}
+while (1) {
+	tailscale_conn conn;
+	if (tailscale_accept(ln, &conn)) {
+		return err(ts);
+	}
+	char buf[2048];
+	while ((ret = read(conn, buf, sizeof(buf))) > 0) {
+		write(1, buf, ret);
+	}
+	close(conn);
+}
+```
+
+For those who have spent some time in our source code, you’ll know
+Tailscale writes almost everything in Go.
+So how did we end up with a C library?
+It’s not widely used, but the standard Go toolchain includes all the
+machinery necessary for producing a C library in the form of
+[build modes](https://pkg.go.dev/cmd/go#hdr-Build_modes).
+
+## Go Build Modes
+
+The two key build modes are `c-archive` and `c-shared`.
+The first creates an object archive that can be linked by a standard
+C toolchain into a binary.
+The second creates a shared object (on linux, the .so file) that can
+be dynamically linked into a binary.
+They are surprisingly easily to use:
+
+    go build -buildmode=c-archive pkg
+
+This produces an object archive, and a generated header for the C
+symbols Go exports.
+
+As part of compilation, cgo can “export” your Go functions to C using
+a `//export` directive.
+This creates a symbol that can be easily called from a C program,
+with cgo taking care of all the work around setting up the Go runtime
+environment necessary to execute the Go code when C invokes the call.
+
+This produces a C interface that works, but like all auto-generated
+interfaces, it is not what you would write by hand.
+So libtailscale has a manually typed out
+[header file](https://github.com/tailscale/libtailscale/blob/main/tailscale.h)
+with documentation, with hand-written shim implementations onto the
+manual interface.
+
+To make it possible to `read(2)` and `write(2)` on a `tailscale_conn`,
+it is one half of a `socketpair(2)`, while the other half is held by
+the userspace TCP/IP implementation.
+
+## Memory Management
+
+Where things get interesting is dealing with managed heap memory.
+Every programming language providing some form of memory safety or
+management imposes (often invisible) rules on allocated heap memory.
+The most dramatic are moving garbage collectors, which can pause some
+fraction of a program at any moment, and move a piece of memory,
+rewriting all pointers into that memory.
+This is extremely common as moving GCs are very CPU-efficient for
+programs that allocate a large amount of short-lived heap memory,
+because they enable very efficient
+[generational memory management](<https://en.wikipedia.org/wiki/Tracing_garbage_collection#Generational_GC_(ephemeral_GC)>).
```

Comment:
```suggestion
[generational memory management](https://en.wikipedia.org/wiki/Tracing_garbage_collection#Generational_GC_(ephemeral_GC)).
```

Reply:
The `<>` was inserted by prettier, which is step in the CI for this website. (I'd never seen it before, apparently it's some markdown trip to deal with the `()` in the link?)

Reply:
Interesting!